### PR TITLE
Replace P Monad with transformer stack

### PR DESCRIPTION
**Pull request by FrontierOtr**
There's really only one purpose for this, which is to embed a "Context" for the parsing into the return type, which indicates which subcommand the parser failed inside, so a better help text can be printed.

### CODE REVIEW

File: src/main/scala/pirate.internal/ParseTraversal.scala [hunk 1]
```diff
@@ -128,8 +127,10 @@ object ParseTraversal {
       }).some
     case CommandParser(sub) =>
       if (sub.name === arg)
-        StateT[P, List[String], A](args =>
-          runParser(SkipOpts, sub.parse, args).map(_.swap)).pure[Option]
+        StateT[P, List[String], A](args => for {
+          _ <- EitherT.right[PWR,ParseError,Unit](().pure[PWR].<++:(arg :: Nil))
```
Comment (FrontierOtr) [inline src/main/scala/pirate.internal/ParseTraversal.scala]: There's probably a more elegant way of doing this.

File: src/main/scala/pirate/Usage.scala [hunk 2]
```diff
@@ -5,11 +5,13 @@ import pirate.internal._
...
-  def print[A](command: Command[A]): String =
-    printWith(command, DefaultUsageMode)
+  def print[A](command: Command[A], context: List[String]): String =
+    printWith(command, context, DefaultUsageMode)
 
-  def printWith[A](command: Command[A], mode: UsageMode): String =
-    Render.infos(command.name, command.description, tree(command.parse), mode)
+  def printWith[A](command: Command[A], context: List[String], mode: UsageMode): String = context match {
+    case Nil => Render.infos(command.name, command.description, tree(command.parse), mode)
+    case x :: _ => printSubWith[A](command, x, mode) // The first entry is the "last" entry, so is the relevant context.
```
Comment (FrontierOtr) [inline src/main/scala/pirate/Usage.scala]: This doesn't walk the tree, so could be wrong if there's name clashes one sub-sub-parsers.

File: src/main/scala/pirate.internal/ParseTraversal.scala [hunk 3]
```diff
@@ -11,6 +11,10 @@ object ParseTraversal {
   case class ShortParsedName(c: Char) extends ParsedName
   case class ParsedWord(name: ParsedName, value: Option[String])
 
+  type P[A] = EitherT[PWR, ParseError, A]
```
Comment (markhibberd) [inline src/main/scala/pirate.internal/ParseTraversal.scala]: I think you will quickly find that you probably want to leave this as a data type wrapping the stack. Gives you somewhere to build out all the specific combinators you will want to make the stack usable in a neater way.

File: src/main/scala/pirate.internal/ParseTraversal.scala (hunk 1 above)
Comment (markhibberd) [inline src/main/scala/pirate.internal/ParseTraversal.scala]: Prob a for comprehension and some better combinators for lifting.

**Comment by markhibberd** [comment]
Stack looks right. I am a bit meh on stacks that deep in scala, because there are some walls where you will find it won't work any more (for example I don't know if you will be able to implement MonadTrans and use liftM generically in this situation). But if it works for now I am fine with it.

**Comment by markhibberd** [comment]
One note though, I think we really need to start adding a much more comprehensive test suite before the internals get any more churn.

**Pull request merged**